File: examples/python/tool_groups.py
# examples/python/tool_groups.py
# Demonstrates tool groups with per-action schemas, validation, and strategies.
# Run: pmcp dev examples/python/tool_groups.py

from typing import Literal, Optional
from protomcp import tool_group, action, ToolResult


@tool_group("db", description="Database operations", strategy="union")
class DatabaseTools:
    """Union strategy: all actions exposed as a single tool with discriminated input."""

    @action("query", description="Run a read-only SQL query", requires=["sql"])
    def query(self, sql: str, limit: int = 100) -> ToolResult:
        return ToolResult(result=f"Executed: {sql} (limit {limit})")

    @action(
        "insert",
        description="Insert a record into a table",
        requires=["table", "data"],
        enum_fields={"table": ["users", "events", "logs"]},
    )
    def insert(self, table: str, data: str) -> ToolResult:
        return ToolResult(result=f"Inserted into {table}: {data}")

    @action("migrate", description="Run a schema migration", requires=["version"])
    def migrate(self, version: str, dry_run: bool = False) -> ToolResult:
        mode = "dry run" if dry_run else "applied"
        return ToolResult(result=f"Migration {version} {mode}")


@tool_group("files", description="File operations", strategy="separate")
class FileTools:
    """Separate strategy: each action becomes its own tool (files.read, files.write, etc.)."""

    @action("read", description="Read a file by path", requires=["path"])
    def read(self, path: str) -> ToolResult:
        return ToolResult(result=f"Contents of {path}")

    @action("write", description="Write content to a file", requires=["path", "content"])
    def write(self, path: str, content: str) -> ToolResult:
        return ToolResult(result=f"Wrote {len(content)} bytes to {path}")

    @action(
        "search",
        description="Search files by pattern",
        requires=["pattern"],
        enum_fields={"scope": ["workspace", "project", "global"]},
    )
    def search(self, pattern: str, scope: Literal["workspace", "project", "global"] = "workspace") -> ToolResult:
        return ToolResult(result=f"Searching '{pattern}' in {scope}")


if __name__ == "__main__":
    from protomcp.runner import run
    run()

File: examples/typescript/tool-groups.ts
// examples/typescript/tool-groups.ts
// Demonstrates tool groups with per-action Zod schemas, validation, and strategies.
// Run: pmcp dev examples/typescript/tool-groups.ts

import { toolGroup, ToolResult } from 'protomcp';
import { z } from 'zod';

// Union strategy (default): single tool "db" with action discriminator
toolGroup({
  name: 'db',
  description: 'Database operations',
  actions: {
    query: {
      description: 'Run a read-only SQL query',
      args: z.object({ sql: z.string(), limit: z.number().default(100) }),
      requires: ['sql'],
      handler({ sql, limit }) {
        return new ToolResult({ result: `Executed: ${sql} (limit ${limit})` });
      },
    },
    insert: {
      description: 'Insert a record into a table',
      args: z.object({ table: z.string(), data: z.string() }),
      requires: ['table', 'data'],
      enumFields: { table: ['users', 'events', 'logs'] },
      handler({ table, data }) {
        return new ToolResult({ result: `Inserted into ${table}: ${data}` });
      },
    },
    migrate: {
      description: 'Run a schema migration',
      args: z.object({ version: z.string(), dry_run: z.boolean().default(false) }),
      requires: ['version'],
      handler({ version, dry_run }) {
        const mode = dry_run ? 'dry run' : 'applied';
        return new ToolResult({ result: `Migration ${version} ${mode}` });
      },
    },
  },
});

// Separate strategy: each action becomes its own tool (files.read, files.write, etc.)
toolGroup({
  name: 'files',
  description: 'File operations',
  strategy: 'separate',
  actions: {
    read: {
      description: 'Read a file by path',
      args: z.object({ path: z.string() }),
      requires: ['path'],
      handler({ path }) {
        return new ToolResult({ result: `Contents of ${path}` });
      },
    },
    write: {
      description: 'Write content to a file',
      args: z.object({ path: z.string(), content: z.string() }),
      requires: ['path', 'content'],
      handler({ path, content }) {
        return new ToolResult({ result: `Wrote ${content.length} bytes to ${path}` });
      },
    },
    search: {
      description: 'Search files by pattern',
      args: z.object({
        pattern: z.string(),
        scope: z.enum(['workspace', 'project', 'global']).default('workspace'),
      }),
      requires: ['pattern'],
      enumFields: { scope: ['workspace', 'project', 'global'] },
      handler({ pattern, scope }) {
        return new ToolResult({ result: `Searching '${pattern}' in ${scope}` });
      },
    },
  },
});
